### DISCUSSION TITLE
DataGrid - Somehow an Achilles' heel?

### DISCUSSION THREAD

**Discussion by marcosgerene** [AvaloniaUI/Avalonia#16235] [General]
Hi folks,

I have been developing with WPF since 2019, but I do not consider myself a master in WPF. I am a background guy and use things like MahApps to make my life easier.

It has also been a while since I started considering (actually, it is already decided now) rewriting the entire UI of my software in Avalonia. The idea of keeping my ViewModels the same and running my application on macOS/Linux/Android Tablets makes me very excited.

Well, just to give you some context, my application is an ERP with something between 200-250 windows. I asked some questions and provided some information/screenshots here: https://github.com/AvaloniaUI/Avalonia/discussions/14188.

I have built some small applications in Avalonia to understand and learn about it, and I had no big problems, but I feel like DataGrid is an Achilles' heel.

Why do I feel like this:
Again, I am not the guy who takes Avalonia and makes big changes; I would use things like Avalonia.Material, Ursa, ActivePro... I want productivity even if I have to pay for a component like ActivePro.Avalonia.

I found everything I wanted for my ERP; even my reports, which are built upon FastReports, have a solution now. But DataGrid does not have (I did not find) solutions, even paid ones.

Things like https://github.com/dotnet/DataGridExtensions are not available. I asked on Ursa about this: https://github.com/irihitech/Ursa.Avalonia/discussions/128.

My whole point is: How true is my feeling? Am I not knowing how to search? Is there any better solution for DataGrids (open-source or not)?

**Comment by robloo** [comment]
DataGrid and table views in general are a problem in this type of framework, and not just for Avalonia.

 * It took forever for WPF's DataGrid to finally be added. This is still the most powerful first-partly DataGrid ever for a XAML platform... but it does have it's quirks. It was NOT open source at the time Avalonia first ported it's DataGrid. Issues with WPF's DataGrid include:
   * It's complexity to maintain -- it's the biggest control to date
   * It's not AOT friendly which is an issue today (Wasn't at the time)
   * It's not MVVM friendly. Columns and a lot of the setup still has to be done in code-behind.
 * Silverlight also needed a DataGrid so wrote another version of it. That Silverlight version is less powerful and has even more quirks (more like actual bugs in this case). However, it's the one other frameworks derived from as it was the first one open sourced. This is the version ported to Avalonia. It is also the version used in UWP's community toolkit at the time.
 * WPF also had something no other XAML framework has (exception noted next): a very light-weight grid/table control. This was the GridView view of a ListView and it was really easy to throw up a table of values (without needing anything advanced).
 * Very recently WinUI added the TableView in the community toolkit which is an attempt at a light-weight control again. This isn't widely used and is a completely new API.
 * Avalonia, in an attempt to fix some pre-existing issues and add new use cases, developed a new TreeDataGrid control. This one is AOT friendly and highly performant (more than DataGrid). However, it's missing a lot of functionality and doesn't really focus on MVVM at all. ~It's still the recommended "table" control to use if you can though.~

So as can be seen this is still an unsolved problem and the whole ecosystem is still searching for a general-purpose solution. This also says nothing about 3rd party DataGrids of which you can find the best implementations. Some 3rd party controls are MVVM friendly and support very complex things in a simple way -- these controls likely took thousands of hours to develop though.

Someday Avalonia might attempt a port of WPF's DataGrid. Someday we might have a light-weight TableView or GridView control as well. ~However, right now use TreeDataGrid if you can first then fall back to the current DataGrid second. Third-party more advanced controls will hopefully step in here too.~

Edit: Comments now reflect the depreciation of TreeDataGrid.

**Reply by marcosgerene** [reply]
@robloo  Thanks, you teach me a lot!

The sad part is I was kind of right, DataGrid is an Achilles' heel, not just for Avalonia but also for Avalonia. =(

**Comment by rabbitism** [comment]
Sorry I didn't reply you in my Ursa repo because I was so addicted to my calendar view... 
In general we do have plan to migrate some heavy DataGrid controls, but we are lack of people (and I'm also busy looking for investors these days).
If you need it urgently, you can check https://eremexcontrols.net/ , it is not free, but afaik, their DataGrid is almost competitive to DevExpress ones.

**Reply by marcosgerene** [reply]
@rabbitism 

In fact you did answer all my questions! I know you have a limited team there, actually I think the work you have already done is amazing.

However, as I said I am looking for some DataGrid components for a while and before eremex you just mentioned I had not seem anything.

Thanks a lot for your effort!

**Comment by timunie** [comment]
Filtering and sorting should be in the VM layer imo. A filter in the control itself would heavily rely on reflection which is not perfect for several reasons imo.

**Reply by marcosgerene** [reply]
@timunie 

I can't discuss this in a deep way, I don't have enough knowledge to this, but...

I think even though it should not be in the "base" component it could be some solution to this, as an "super DataGrid" developed by a company as ActivePro (or Eremex as @rabbitism has mentioned before). It is hard to find even paid solutions to this.

The idea o this "lack of solution" is an Achilles' heel, that was my point.

**Reply by timunie** [reply]
Sure, it would be great to have a huge 3rd party eco system available and I'm sure it will grow further. But right now,

**Reply by thevortexcloud** [reply]
One less than ideal solution is to use XPF hybrid which will let you run third party WPF controls inside Avalonia on most of the platforms Avalonia supports. The Teleirk grid stuff for example apparently works.

https://avaloniaui.net/xpf/packages/telerik

You could also probably use MAUI controls with MAUI hybrid stuff.

https://github.com/AvaloniaUI/AvaloniaMauiHybrid

Although MAUI does not support the same platforms as Avalonia, so that may not work for you.

**Reply by marcosgerene** [reply]
@thevortexcloud 
I think XPF is a great solution, but the idea of have a new project with something "adapted" from WPF does not sound well to me.

I have consired XPF to my actual WPF project, but I have decided to rewrite the entire UI, I will do this using only Avalonia things. My first idea was to use only open-source solutions, but I face a lack of solutions on DataGrid.

By now I will try the new TreeDataGrid and maybe adapt some functions if needed.

**Reply by JakkuSakura** [reply]
FYI, I've implemented basic filtering in TreeDataGrid. It's currently in PR

**Comment by Csemakula** [comment]
I may be mistaken but there does seem to be a sort of DataGrid that Avalonia do call DataGrid. I am reviewing Avalonia and I did see something of the sort in their documentation.

https://docs.avaloniaui.net/docs/reference/controls/datagrid/

I am thinking of porting an app I created for Windows to MacOs so I was looking for something like this since my app does use quite a lot of grids and stuff. Without grids the port would be almost useless as my users would not be able to use it. Now my next problem is finding a way to take that grid to Excel. Don't know when it was added in though...

**Reply by thevortexcloud** [reply]
The DataGrid in Avalonia works fine. It's not super advanced, but it's "good enough" for most use cases. From memory I seem to recall reading that it's just a port of the Silverlight DataGrid? But I may be wrong.

> Now my next problem is finding a way to take that grid to Excel

There are many libraries for generating Excel files. It's also very easy to generate CSV files yourself if need be. But I don't see Avalonia integrating a generic grid export feature any time soon.

**Comment by DevAM-Tools** [comment]
> My whole point is: How true is my feeling? Am I not knowing how to search? Is there any better solution for DataGrids (open-source or not)?

Unfortunately [TreeDataGrid](https://github.com/AvaloniaUI/Avalonia.Controls.TreeDataGrid) has steep learning curve but once you got it it is superior to a `DataGrid`. (feature wise and performance wise) It is also customisable to a high degree.

**Reply by thevortexcloud** [reply]
Tree Data Grid is also being deprecated now.

https://github.com/AvaloniaUI/Avalonia/discussions/16997#discussioncomment-10717267

**Reply by robloo** [reply]
Yea, my original comments need to be amended to take that into account.

**Reply by MikeCodesDotNET** [reply]
I need to clarify our current thinking on TreeDataGrid (TDG). 

Right now, everything will stay the same until at least Q3 2025. The plan is to deprecate the OSS version and ship an improved version as part of Avalonia Accelerate. This version will not be dual-licensed as initially intended, but we will point users to a community fork of the old TDG for those who want to use a free version. 

This plan may still change, but we'll keep everyone updated on any changes to the strategy.

**Reply by cirrusone** [reply]
Are you willing to reveal rough launch date of Avalonia Accelerate for phase 1? I'm in need of an auth web login and Web Dialog sounds like it may help.

**Reply by MikeCodesDotNET** [reply]
> Are you willing to reveal rough launch date of Avalonia Accelerate for phase 1? 

Not just yet. The plan is to have a private beta release before Christmas for our existing customers, but we're not 100% fixed on that.

**Reply by DevAM-Tools** [reply]
I want to highlight that having a powerful high performant DataGrid-like Control is essential for an UI framework. (imo)

I would really like to see TreeDataGrid be better integrated into the development of Avalonia. In the form that it is understood to be an integral part of Avalonia.

**Reply by MikeCodesDotNET** [reply]
> I would really like to see TreeDataGrid be better integrated into the development of Avalonia. In the form that it is understood to be an integral part of Avalonia.

We would love to see this as well, which is why we're looking to move the TreeDataGrid into our commercial offering. This way we can better invest in its future.

**Reply by marcosgerene** [reply]
@MikeCodesDotNET 

I had in the past plans to use XPF, but looking to my small business needs and XPF costs, I have decided to migrate from WPF to Avalonia in an old-style way, but it will take some time (we are talking of an ERP, around 250 windows).

During my studies and searches on Avalonia, I missed two things on DataGrid. Filtering and Paging.

I wanted something as a "DataGrid Plus", I was not looking to some ultra-powerfull tool. I couldn't find it, paid or not. 
Free/OpenSource does not implement better DataGrid. The single paid tool I found was too expensive.

**Reply by MikeCodesDotNET** [reply]
> The single paid tool I found was too expensive.

Out of interest. What is your budget?

**Reply by marcosgerene** [reply]
@MikeCodesDotNET 

The "Avalonia Pro" price from Actipro is fine to me. But I don't need their controls and the control I need they don't have. 

![image](https://github.com/user-attachments/assets/0f2cb23a-c6cb-46d7-bbcc-b6a3b6ce72ae)

**Reply by jm-abellan** [reply]
...yet
They really trust Avalonia is to be the next cross-platform framework and they are working hard on new controls as far as I know
I am one of their customers and I must say that the quality of their job is much higher than low is the price

**Comment by mateli** [comment]
I think that the smart thing to do would be to make controls based on Skia that works with multiple toolkits. There are some graphing controls that are good examples of this.

That is what I would do if I were ActivePro. I would make a "host" control for WPF, Avalonia etc and then controls that run inside that host control. Then supporting new framework becomes very easy.

Furthermore I agree that ActivePro contains "everything but what I need". Making prettier buttons or something like that means very little to me. Having controls that can display data in flexible ways is what I want. In my opinion Swing is still the higher end of the toolkit scale when it comes to that.

**Reply by JakkuSakura** [reply]
It's similar to a "headless" approach like tanstack table

**Comment by JakkuSakura** [comment]
Hi fellows, I spent some time migrating tanstack to Avalonia
https://github.com/JakkuSakura/SaGrid
<img width="1400" height="928" alt="image" src="https://github.com/user-attachments/assets/97b7ad18-ba99-479b-af39-583dbf335118" />

I also opened a [PR](https://github.com/AvaloniaUI/Avalonia.Controls.TreeDataGrid/pull/382) to TreeDataGrid, but I don't like its complicated source code

**Comment by DevAM-Tools** [comment]
I am still very concerned regarding the future of datagrids for Avalonia. When all free existing solutions (DataGrid and TreeDataGrid) are depricated there is a serious risk that they will be outdated very soon. Especially when we get to Avalonia 12.
In my opinion a datagrid control is an essential element of an UI framework. A UI framework that does not offer a datagrid is not competitive from my point of view. Especially when it targets the desktop.

If I reflect my major uses cases then I need both a high performant (more than 100 million rows) TreeView and a tabular view with customizable Cells and headers.

I also require filtering, grouping or sorting in some scenarios. But in my opinion the UI and features for these are very specific to the use case. Therefor I would agree to opinion written above: This should be done in the ViewModel with an custom UI in the data grid headers.

For now I am happy with TreeDataGrid as it enables these use cases. So please don't deprecate TreeDataGrid and keep it alive as kind of free basic datagrid like control for Avalonia.

**Reply by MikeCodesDotNET** [reply]
You're absolutely right that DataGrids are essential for desktop applications and we don't disagree. 

The challenge is that maintaining a production-quality DataGrid requires continuous, specialised effort. As we've seen from the complete lack of response to our [call for maintainers](https://github.com/AvaloniaUI/Avalonia/discussions/19653) (still zero volunteers after 2+ weeks), the community wants to use these controls but has little interest in helping to meaningfully build and maintain them.

The TreeDataGrid contribution history tells the same story with @grokys having done virtually all the heavy lifting.

The existing FOSS TreeDataGrid code will remain available under its current license. Anyone is free to fork and maintain it if they wish. For those who need professional support and continued development, Accelerate will provide that path forward with an actively maintained and enhanced version.

With v12, we will be introducing a free, read-only control for displaying tabular data. It won't have the full complexity of a DataGrid (no editing, etc.), but it will provide a way to display data in a grid format entirely for free. The goal with this is to ensure that basic tabular display (which many applications need) remains accessible to everyone.

Your use cases (high-performance TreeView, customizable cells/headers) represent exactly the kind of complex requirements that need sustainable, professional-grade support. These advanced scenarios will be best served through either Accelerate or third-party solutions like Eremex.

I understand this isn't ideal from your perspective, but it ensures these critical controls actually get the maintenance and development they deserve rather than slowly becoming outdated due to lack of resources.
